Source/Particles/MultiParticleContainer.cpp

```diff
@@ -948,4 +973,136 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
         m_shr_p_bw_engine->init_lookup_tables_from_raw_data(table_data);
     }
 }
+
+void MultiParticleContainer::doQedEvents()
+{
+    WARPX_PROFILE("MPC::doQedEvents");
+
+    doQedBreitWheeler();
+    doQedQuantumSync();
+}
+
+void MultiParticleContainer::doQedBreitWheeler()
+{
+    WARPX_PROFILE("MPC::doQedEvents::doQedBreitWheeler");
+
+    // Loop over all species.
+    // Photons undergoing Breit Wheeler process create electrons
+    // in pc_product_ele and positrons in pc_product_pos
+
+    for (auto& pc_source : allcontainers){
+        if(!pc_source->has_breit_wheeler()) continue;
+
+        // Get product species
+        auto& pc_product_ele =
+            allcontainers[pc_source->m_qed_breit_wheeler_ele_product];
+        auto& pc_product_pos =
+            allcontainers[pc_source->m_qed_breit_wheeler_pos_product];
+
+        SmartCopyFactory copy_factory_ele(*pc_source, *pc_product_ele);
+        SmartCopyFactory copy_factory_pos(*pc_source, *pc_product_pos);
+        auto phys_pc_ptr = static_cast<PhysicalParticleContainer*>(pc_source.get());
+
+        const auto Filter  = phys_pc_ptr->getPairGenerationFilterFunc();
+        const auto CopyEle = copy_factory_ele.getSmartCopy();
+        const auto CopyPos = copy_factory_pos.getSmartCopy();
+
+        const auto pair_gen_functor = m_shr_p_bw_engine->build_pair_functor();
+        auto Transform = PairGenerationTransformFunc(pair_gen_functor);
+
+        pc_source ->defineAllParticleTiles();
+        pc_product_pos->defineAllParticleTiles();
+        pc_product_ele->defineAllParticleTiles();
+
+        for (int lev = 0; lev <= pc_source->finestLevel(); ++lev)
+        {
+            auto info = getMFItInfo(*pc_source, *pc_product_ele, *pc_product_pos);
+
+#ifdef _OPENMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+            for (MFIter mfi = pc_source->MakeMFIter(lev, info); mfi.isValid(); ++mfi)
+            {
+                auto& src_tile = pc_source ->ParticlesAt(lev, mfi);
```

Comment:
```suggestion
                const auto& src_tile = pc_source->ParticlesAt(lev, mfi);
```
It may be worth trying putting a bunch of `const` here and there.

Reply:
Unfortunately it seems that `src_tile` and `dst_tile` cannot be `const` due to the fact that they are passed to `filterCopyTransformParticles`. However, I've added a bunch of `const` as requested!